File: scripts/verify-abstract-algebra.py
"""Independent permutation/matrix/enumeration oracles for actual production exports."""

from pathlib import Path
from fractions import Fraction as F
from itertools import permutations, product
from datetime import datetime, timezone
import contextlib
import io
import json
import math
import sympy as sp

ROOT = Path(__file__).resolve().parents[1]
DIRECTORY = ROOT / 'scratch/abstract-algebra-verification'
data = json.loads((DIRECTORY/'payload.json').read_text())
counts = {}
exact_matrix = lambda rows: sp.Matrix([[sp.Rational(str(v)) for v in row] for row in rows])
exact_vector = lambda values: sp.Matrix([sp.Rational(str(v)) for v in values])


def check(condition, label):
    assert condition, label
    counts[label] = counts.get(label, 0) + 1


namespaces = {}
for example in data['examples']:
    namespace, output = {}, io.StringIO()
    with contextlib.redirect_stdout(output):
        exec(compile(example['code'], example['id'], 'exec'), namespace)
    check(output.getvalue().strip() == example['expected'], 'actual_program_stdout')
    namespaces[example['id']] = namespace

# Independent group model: exact geometric matrices, not the encoded pair law.
R, S = sp.Matrix([[0,-1],[1,0]]), sp.diag(1,-1)
matrices = [R**k * S**b for b in range(2) for k in range(4)]
vertices = [sp.Matrix([1,0]),sp.Matrix([0,1]),sp.Matrix([-1,0]),sp.Matrix([0,-1])]
perms = [tuple(vertices.index(matrix*point) for point in vertices) for matrix in matrices]
Ps = [sp.Matrix(4,4,lambda j,i:int(p[i]==j)) for p in perms]
def comp(g,h): return matrices.index(matrices[g]*matrices[h])
def move(g,state): return tuple(state[perms[g].index(j)] for j in range(4))

distance = lambda i,j: (vertices[i]-vertices[j]).dot(vertices[i]-vertices[j])
geometric = {p for p in permutations(range(4)) if all(distance(i,j)==distance(p[i],p[j]) for i,j in product(range(4),repeat=2))}
check(geometric==set(perms), 'independent_distance_symmetries')
for g in range(8):
    check(data['permutations'][g]==list(perms[g]), 'geometric_vertex_maps')
    check(data['matrices'][g]==matrices[g].tolist(), 'geometric_representation')
    check(matrices[data['inverse'][g]]==matrices[g].inv(), 'exact_inverse')
for row in data['compositions']:
    g,h,v=row['g'],row['h'],row['vertex']
    check(row['product']==comp(g,h) and row['reversed']==comp(h,g), 'actual_composition_matrix_oracle')
    check(row['firstRoute']==[v,perms[h][v],perms[comp(g,h)][v]], 'actual_tracked_route')
    check(row['firstStates'][-1]==list(move(comp(g,h),'ABCD')), 'label_source_destination')

# Solve subgroup membership by independent permutation closure, then derive cosets.
valid_subgroups=[]
for row in data['subgroups']:
    H=set(row['subset'])
    valid=bool(H) and 0 in H and all(comp(g,h) in H for g in H for h in H)
    check((row['cosets'] is not None)==valid, 'all_256_subset_classifications')
    generated={0}
    while True:
        expanded=generated|{comp(g,h) for g in H for h in generated}
        if generated==expanded: break
        generated=expanded
    check(set(row['generated'])==generated, 'changed_generated_subgroups')
    if valid:
        valid_subgroups.append(H)
        expected={frozenset(comp(g,h) for h in H) for g in range(8)}
        actual={frozenset(c['members']) for c in row['cosets']}
        check(actual==expected, 'all_subgroup_cosets')
check(len(valid_subgroups)==10, 'D4_subgroup_count')

for row in data['orbits']:
    x=tuple(row['colors']); group=row['group']
    expected={move(g,x) for g in group}
    check({tuple(s['colors']) for s in row['states']}==expected, 'all_ternary_orbits')
    stabilizer={g for g in group if move(g,x)==x}
    check(set(row['stabilizer'])==stabilizer, 'all_ternary_stabilizers')
    check(len(expected)*len(stabilizer)==len(group), 'orbit_stabilizer_identity')
    for state in row['states']:
        check(set(state['transformations'])=={g for g in group if move(g,x)==tuple(state['colors'])}, 'exact_reaching_transformations')
for row in data['fixed']:
    states=list(product(range(row['colorCount']),repeat=4)); group=row['group']
    for entry in row['rows']:
        check(entry['fixed']==sum(move(entry['g'],state)==state for state in states), 'fixed_counts_by_enumeration')
    classes={min(move(g,state) for g in group) for state in states}
    check(row['orbitCount']==len(classes), 'Burnside_vs_full_enumeration')

# Matrix orbit averages directly pool three geometric relation classes, rather than
# computing conjugates again. Also independently solve the commutant equations.
symbols=sp.symbols('w0:16'); symbolic=sp.Matrix(4,4,symbols)
equations=[]
for p in [Ps[1],Ps[4]]: equations.extend(list(symbolic*p-p*symbolic))
constraint,_=sp.linear_eq_to_matrix(equations,symbols)
check(len(constraint.nullspace())==3,'independent_commutant_dimension')
for row in data['averages']:
    W=sp.Matrix([[sp.Rational(str(v)) for v in line] for line in row['matrix']])
    coefficient={kind:sum(W[i,j] for i in range(4) for j in range(4) if (min((j-i)%4,(i-j)%4))==kind)/sum(min((j-i)%4,(i-j)%4)==kind for i in range(4) for j in range(4)) for kind in range(3)}
    full=sp.Matrix(4,4,lambda i,j:coefficient[min((j-i)%4,(i-j)%4)])
    offsets={k:sum(W[i,(i+k)%4] for i in range(4))/4 for k in range(4)}
    rotation=sp.Matrix(4,4,lambda i,j:offsets[(j-i)%4])
    check(exact_matrix(row['full'])==full,'changed_average_relation_oracle')
    check(exact_matrix(row['rotations'])==rotation,'changed_rotation_offset_oracle')
    T=sp.Matrix(4,4,lambda i,j:[2,3,5,3][(j-i)%4])
    norm=lambda A:sum(v*v for v in A)
    check(norm(W-T)==norm(W-full)+norm(full-T),'exact_projection_pythagoras')
    actual=namespaces['averaging']['averaged']([[F(str(v)) for v in line] for line in row['matrix']])
    check(sp.Matrix(actual)==full,'actual_native_changed_averages')

for row in data['equivariance']:
    W=exact_matrix(row['matrix']); x=exact_vector(row['values']); p=Ps[row['g']]
    check(W*p*x==exact_vector(row['transformThenMap']),'actual_two_route_matrix_oracle')
    check(p*W*x==exact_vector(row['mapThenTransform']),'actual_two_route_matrix_oracle')
    check(float(max(abs(v) for v in W*p*x-p*W*x))==row['inputSpecificDefect'],'actual_input_defect')
    defects=[max(abs(v) for v in W*q-q*W) for q in Ps]
    check(float(max(defects))==row['allInputDefect'],'actual_full_matrix_certificate')
    check(float(sum(W*x)/4)==row['poolBefore'] and float(sum(W*p*x)/4)==row['poolAfter'],'actual_input_to_pooled_output')

for row in data['modular']:
    n,a,b=row['modulus'],row['multiplier'],row['target']
    solutions=[x for x in range(n) if (a*x-b)%n==0]
    check(row['solutions']==solutions,'all_bounded_modular_solutions')
    check(row['injective']==(math.gcd(a,n)==1),'unit_gcd_oracle')

# Execute the specifically changed independent practice capstone.
capstone=next(example for example in data['examples'] if example['id']=='capstone')['code']
assert '[2, 2, 4, 8]' in capstone and 'x = (1, 2, 4, 8)' in capstone
changed=capstone.replace('[2, 2, 4, 8]','[6, 2, 4, 8]').replace('x = (1, 2, 4, 8)','x = (-1, 0, 2, 3)')
namespace,output={},io.StringIO()
with contextlib.redirect_stdout(output): exec(changed,namespace)
check(namespace['matvec'](namespace['A'],namespace['x'])==(F(133,4),F(67,2),F(37),F(149,4)),'actual_changed_capstone_output')
check(namespace['means']=={F(141,4)},'actual_changed_capstone_invariant_mean')
check(move(4,(0,0,1,2)) not in {move(g,(0,0,1,2)) for g in range(4)},'ternary_chiral_practice')
check(namespaces['modular']['solve_congruence'](3,4,8)==[4],'changed_modular_practice')
result={'checkedAt':datetime.now(timezone.utc).isoformat(),'status':'passed','productionSources':data['productionSources'],'checks':counts,'invalidModelCases':data['rejectionCount'],'changedCapstoneStdout':output.getvalue().strip(),'environment':{'sympy':sp.__version__},'limits':'Finite exact arithmetic, enumeration and separately formulated matrix oracles check actual production exports; mathematical proofs and actual-font browser reading are recorded separately.'}
(DIRECTORY/'results.json').write_text(json.dumps(result,indent=2)+'\n',encoding='utf-8')
print(json.dumps(result,indent=2))

File: scripts/verify-abstract-algebra.mjs
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
import assert from 'node:assert/strict';
import { spawnSync } from 'node:child_process';
import * as model from '../src/learn/data/abstract-algebra-models.js';
import { abstractAlgebraExamples } from '../src/learn/data/abstract-algebra-examples.js';

const directory = 'scratch/abstract-algebra-verification';
fs.mkdirSync(directory, { recursive: true });
const sourcePaths = [
  'src/learn/data/topics/abstract-algebra-groups-symmetry-actions.jsx',
  'src/learn/data/abstract-algebra-models.js',
  'src/learn/data/abstract-algebra-examples.js',
  'src/learn/components/lesson-labs/AbstractAlgebraLabs.jsx',
  'src/learn/components/lesson-labs/abstract-algebra-labs.css',
  'src/learn/data/curriculum/blueprints/abstract-algebra-groups-symmetry-actions.js',
];
const data = {
  productionSources: sourcePaths.map(source => ({ path: source, sha256: crypto.createHash('sha256').update(fs.readFileSync(source)).digest('hex') })),
  examples: abstractAlgebraExamples,
  compositions: [], permutations: [], matrices: [], inverse: [], subgroups: [], orbits: [], fixed: [], averages: [], equivariance: [], modular: [],
};
for (let g = 0; g < 8; g += 1) {
  data.permutations.push(model.squarePermutation(g));
  data.matrices.push(model.squareMatrix(g));
  data.inverse.push(model.inverseSquare(g));
  for (let h = 0; h < 8; h += 1) data.compositions.push(model.compositionState(g, h, (g + h) % 4));
}
for (let mask = 0; mask < 256; mask += 1) {
  const subset = model.squareGroup.filter(g => mask & (1 << g));
  let cosets = null;
  try { cosets = model.leftCosets(subset); } catch (error) { assert(error instanceof RangeError || error instanceof TypeError); }
  data.subgroups.push({ subset, generated: model.generatedSubgroup(subset), cosets });
}
for (let index = 0; index < 81; index += 1) {
  const colors = [0, 1, 2, 3].map(power => Math.floor(index / (3 ** power)) % 3);
  for (const only of [false, true]) data.orbits.push(model.orbitState(colors, only));
}
for (const q of [2, 3, 4]) for (const only of [false, true]) data.fixed.push(model.fixedColoringCount(q, only));
for (let seed = 0; seed < 32; seed += 1) {
  const matrix = Array.from({ length: 4 }, (_, i) => Array.from({ length: 4 }, (_, j) => ((17 * seed + 7 * i + 11 * j + i * j) % 33 - 16) / 4));
  data.averages.push({ matrix, full: model.averageSquareMap(matrix), rotations: model.averageSquareMap(matrix, true) });
  assert.deepEqual(model.averageSquareMap(model.averageSquareMap(matrix)), model.averageSquareMap(matrix));
  const values = [0, 1, 2, 3].map(i => ((seed * 13 + i * 7) % 41 - 20) / 4);
  const weights = [((seed % 9) - 4) / 4, (((seed * 3) % 9) - 4) / 4, (((seed * 5) % 9) - 4) / 4];
  for (const mode of ['raw', 'rotations', 'averaged', 'tied', 'shift']) {
    for (let g = 0; g < 8; g += 1) data.equivariance.push(model.equivarianceState(values, mode, g, weights));
  }
}
for (let n = 2; n <= 12; n += 1) for (let a = 0; a < n; a += 1) for (let b = 0; b < n; b += 1) data.modular.push(model.modularState(n, a, b));
let rejectionCount = 0;
for (const action of [
  () => model.composeSquare(-1, 0), () => model.composeSquare(0, 8), () => model.squarePermutation(NaN),
  () => model.orbitState([0, , 0, 1]), () => model.orbitState([0, 0, 0, 3]), () => model.orbitState([0, 0, 0, 1], 'yes'),
  () => model.leftCosets([]), () => model.leftCosets([0, 1]), () => model.generatedSubgroup([, 1]),
  () => model.fixedColoringCount(1), () => model.fixedColoringCount(2.5),
  () => model.averageSquareMap([[1], [2], [3], [4]]),
  () => model.averageSquareMap(Array.from({ length: 4 }, () => [1, 2, 3, Infinity])),
  () => model.equivarianceState([1, 2, 3, 4], 'toString'), () => model.equivarianceState([1, 2, , 4]),
  () => model.equivarianceState([1, 2, 3, 0.1]), () => model.parseSensorReadings('1,2,,4'),
  () => model.parseSensorReadings('1,2,3,1e-100'), () => model.parseSensorReadings('1,2,3,21'),
  () => model.modularState(5, 5, 1), () => model.modularState(6, 2, 1.5),
]) { assert.throws(action); rejectionCount += 1; }
assert.deepEqual(model.parseSensorReadings(' -1.25, +2, .5, 0 '), [-1.25, 2, 0.5, 0]);
assert.equal(model.cosetProductState(false, 0).sameOutput, true);
assert.equal(model.cosetProductState(false, 1).sameOutput, false);
for (let index = 0; index < 4; index += 1) assert(model.cosetProductState(true, index).sameOutput);
data.rejectionCount = rejectionCount;
fs.writeFileSync(`${directory}/payload.json`, JSON.stringify(data));
const run = spawnSync(path.resolve('scratch/lesson-tools/Scripts/python.exe'), ['scripts/verify-abstract-algebra.py'], { encoding: 'utf8', env: { ...process.env, PYTHONIOENCODING: 'utf-8' } });
process.stdout.write(run.stdout);
process.stderr.write(run.stderr);
process.exitCode = run.status ?? 1;
